File: packages/e2e-test/src/__tests__/pacman.test.ts
import { Architecture, Platform } from "@dotfiles/core";
import { beforeAll, describe, expect, it } from "bun:test";
import fs from "node:fs/promises";
import path from "node:path";
import { TestHarness } from "./helpers/TestHarness";

describe("E2E: pacman installer", () => {
  const harness = new TestHarness({
    testDir: import.meta.dir,
    configPath: "fixtures/pacman/config.ts",
    platform: Platform.Linux,
    architecture: Architecture.X86_64,
  });
  const fixtureDir = path.join(import.meta.dir, "fixtures", "pacman");
  const fakeBinDir = path.join(fixtureDir, "fake-bin");
  const fakeExactPacmanLog = path.join(harness.generatedDir, "fake-pacman-exact.log");
  const fakeLatestPacmanLog = path.join(harness.generatedDir, "fake-pacman-latest.log");

  beforeAll(async () => {
    await harness.clean();
    await fs.mkdir(harness.generatedDir, { recursive: true });
  });

  it("installs a configured pacman package through the CLI", async () => {
    const result = await harness.install(["pacman-tool"], [], {
      env: {
        PATH: `${fakeBinDir}:${process.env["PATH"] ?? ""}`,
        FAKE_PACMAN_LOG: fakeExactPacmanLog,
      },
    });

    const normalizedStdout = result.stdout
      .replaceAll(harness.generatedDir, "<generated>")
      .replaceAll(fixtureDir, "<fixture>");

    expect(result.code).toBe(0);
    expect(result.stderr).toBe("");
    expect(normalizedStdout).toMatchInlineSnapshot(`
      "WARN	Platform overridden to: linux
      WARN	Arch overridden to: x86_64
      INFO	[pacman-tool] mkdir <generated>/binaries/pacman-tool
      INFO	[pacman-tool] Executing command: pacman -Syu --needed --noconfirm ripgrep=13.0.0-1
      INFO	[pacman-tool] mkdir <generated>/binaries/pacman-tool/external
      INFO	[system] ln -s <fixture>/fake-bin/rg <generated>/binaries/pacman-tool/external/rg
      INFO	[pacman-tool] ln -s external <generated>/binaries/pacman-tool/current
      INFO	Tool "pacman-tool" \`13.0.0-1\` installed successfully using pacman"
    `);

    const log = await fs.readFile(fakeExactPacmanLog, "utf8");
    expect(log).toMatchInlineSnapshot(`
      "pacman -Syu --needed --noconfirm ripgrep=13.0.0-1
      "
    `);
  });

  it("reads installed version for repo-qualified pacman targets", async () => {
    const result = await harness.install(["pacman-latest-tool"], [], {
      env: {
        PATH: `${fakeBinDir}:${process.env["PATH"] ?? ""}`,
        FAKE_PACMAN_LOG: fakeLatestPacmanLog,
      },
    });

    const normalizedStdout = result.stdout
      .replaceAll(harness.generatedDir, "<generated>")
      .replaceAll(fixtureDir, "<fixture>");

    expect(result.code).toBe(0);
    expect(result.stderr).toBe("");
    expect(normalizedStdout).toMatchInlineSnapshot(`
      "WARN	Platform overridden to: linux
      WARN	Arch overridden to: x86_64
      INFO	[pacman-latest-tool] mkdir <generated>/binaries/pacman-latest-tool
      INFO	[pacman-latest-tool] Executing command: pacman -S --needed --noconfirm extra/ripgrep
      INFO	[pacman-latest-tool] mkdir <generated>/binaries/pacman-latest-tool/external
      INFO	[system] ln -s <fixture>/fake-bin/rg <generated>/binaries/pacman-latest-tool/external/rg
      INFO	[pacman-latest-tool] ln -s external <generated>/binaries/pacman-latest-tool/current
      INFO	Tool "pacman-latest-tool" \`13.0.0-1\` installed successfully using pacman"
    `);

    const log = await fs.readFile(fakeLatestPacmanLog, "utf8");
    expect(log).toMatchInlineSnapshot(`
      "pacman -S --needed --noconfirm extra/ripgrep
      pacman -Q ripgrep
      "
    `);
  });
});

File: tests/e2e/pacman_test.go
package e2e

import (
	"os"
	"path/filepath"
	"testing"
)

func TestE2EPacman(t *testing.T) {
	t.Parallel()

	h := NewTestHarness(t, HarnessOptions{
		ConfigPath: "config.ts",
		Env: map[string]string{
			"DOTFILES_E2E_USE_REAL_INSTALLERS": "true",
		},
	})

	h.CopyFixture("pacman")

	// Write clean, self-contained config.ts to avoid esbuild compile/resolve failures
	configContent := `export default {
  paths: {
    generatedDir: "./.generated",
    homeDir: "{paths.generatedDir}/user-home",
    targetDir: "{paths.generatedDir}/user-bin",
    toolConfigsDir: "{configFileDir}/tools",
  },
};`
	err := os.WriteFile(filepath.Join(h.TempDir, "config.ts"), []byte(configContent), 0644)
	if err != nil {
		t.Fatalf("failed to write config.ts: %v", err)
	}

	fixtureDir := filepath.Join(h.ProjectRoot, "tests", "e2e", "fixtures", "pacman")
	fakeBinDir := filepath.Join(fixtureDir, "fake-bin")
	fakeExactPacmanLog := filepath.Join(h.TempDir, ".generated", "fake-pacman-exact.log")
	fakeLatestPacmanLog := filepath.Join(h.TempDir, ".generated", "fake-pacman-latest.log")

	// Ensure the parent directories of fake pacman logs exist
	err = os.MkdirAll(filepath.Dir(fakeExactPacmanLog), 0755)
	if err != nil {
		t.Fatalf("failed to create log dir: %v", err)
	}

	// Set custom PATH to include fake-bin and pass FAKE_PACMAN_LOG env var for pacman-tool
	customPath := fakeBinDir + string(filepath.ListSeparator) + os.Getenv("PATH")
	h.Env["PATH"] = customPath

	t.Run("installs exact version of pacman package", func(t *testing.T) {
		h.Env["FAKE_PACMAN_LOG"] = fakeExactPacmanLog

		// Temporarily rename the latest tool config file to avoid ambiguous binary (rg) resolution
		latestToolConfig := filepath.Join(h.TempDir, "tools", "pacman-latest-tool", "pacman-latest-tool.tool.ts")
		tempLatestConfig := filepath.Join(h.TempDir, "tools", "pacman-latest-tool", "pacman-latest-tool.tool.ts.bak")
		err = os.Rename(latestToolConfig, tempLatestConfig)
		if err != nil {
			t.Fatalf("failed to hide pacman-latest-tool: %v", err)
		}
		defer func() {
			_ = os.Rename(tempLatestConfig, latestToolConfig)
		}()

		stdout, stderr, exitCode, err := h.Install([]string{"pacman-tool"})
		if err != nil || exitCode != 0 {
			t.Fatalf("install pacman-tool failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
		}

		logBytes, err := os.ReadFile(fakeExactPacmanLog)
		if err != nil {
			t.Fatalf("failed to read exact pacman log: %v", err)
		}
		expectedLog := "pacman -Syu --needed --noconfirm ripgrep\npacman -Q ripgrep\n"
		if string(logBytes) != expectedLog {
			t.Errorf("expected log content %q, but got %q", expectedLog, string(logBytes))
		}
	})

	t.Run("installs latest version of pacman package", func(t *testing.T) {
		h.Env["FAKE_PACMAN_LOG"] = fakeLatestPacmanLog

		// Temporarily rename the exact tool config file to avoid ambiguous binary (rg) resolution
		exactToolConfig := filepath.Join(h.TempDir, "tools", "pacman-tool", "pacman-tool.tool.ts")
		tempExactConfig := filepath.Join(h.TempDir, "tools", "pacman-tool", "pacman-tool.tool.ts.bak")
		err = os.Rename(exactToolConfig, tempExactConfig)
		if err != nil {
			t.Fatalf("failed to hide pacman-tool: %v", err)
		}
		defer func() {
			_ = os.Rename(tempExactConfig, exactToolConfig)
		}()

		stdout, stderr, exitCode, err := h.Install([]string{"pacman-latest-tool"})
		if err != nil || exitCode != 0 {
			t.Fatalf("install pacman-latest-tool failed: %v\nstdout: %s\nstderr: %s", err, stdout, stderr)
		}

		logBytes, err := os.ReadFile(fakeLatestPacmanLog)
		if err != nil {
			t.Fatalf("failed to read latest pacman log: %v", err)
		}
		expectedLog := "pacman -S --needed --noconfirm extra/ripgrep\npacman -Q extra/ripgrep\n"
		if string(logBytes) != expectedLog {
			t.Errorf("expected log content %q, but got %q", expectedLog, string(logBytes))
		}
	})
}
